File: internal/i18n/extract/scope.go
// SPDX-FileCopyrightText: 2026-present Brian Wang <wangbuke@gmail.com>
// SPDX-License-Identifier: LGPL-3.0-or-later

package extract

import (
	"path/filepath"
	"strings"
)

// FormatScope builds Scope = path[@location], matching modules/core/service/i18n/scope.ts.
func FormatScope(path string, location string) string {
	p := strings.TrimSpace(path)
	loc := strings.TrimSpace(location)
	if p == "" {
		return loc
	}
	if loc == "" {
		return p
	}
	return p + "@" + loc
}

// ResolveI18nScope mirrors runtime resolveI18nScope:
// explicit scope → withI18nScope stack top → path[@location].
func ResolveI18nScope(manualScope string, scopeStack []string, path string, location string) string {
	if manual := strings.TrimSpace(manualScope); manual != "" {
		return manual
	}
	if len(scopeStack) > 0 {
		return scopeStack[len(scopeStack)-1]
	}
	return FormatScope(path, location)
}

// ScopePathFromRelPath converts a module-relative file path into the auto Scope path
// (no extension, forward slashes), e.g. web/pages/Login.vue → web/pages/Login.
func ScopePathFromRelPath(relPath string) string {
	relPath = filepath.ToSlash(strings.TrimSpace(relPath))
	relPath = strings.TrimPrefix(relPath, "./")
	ext := filepath.Ext(relPath)
	if ext != "" {
		relPath = strings.TrimSuffix(relPath, ext)
	}
	return relPath
}

File: modules/core/service/i18n/scope.ts
// SPDX-FileCopyrightText: 2026-present Brian Wang <wangbuke@gmail.com>
// SPDX-License-Identifier: Apache-2.0

/**
 * Scope = path[@location] | manual scope override.
 * See terminology-i18n-design.md §5.3.
 */

const scopeStack: string[] = [];

export function formatScope(path: string, location?: string): string {
  const p = (path || '').trim();
  const loc = (location || '').trim();
  if (!p) {
    return loc || '';
  }
  if (!loc) {
    return p;
  }
  return `${p}@${loc}`;
}

export function withI18nScope<T>(scope: string, fn: () => T): T {
  scopeStack.push(scope);
  try {
    return fn();
  } finally {
    scopeStack.pop();
  }
}

export type ResolveI18nScopeOptions = {
  /** Manual scope override (wins over stack / path). */
  scope?: string;
  /** Source file path used for auto scope. */
  path?: string;
  /** Optional location suffix for auto scope. */
  location?: string;
};

/**
 * Resolve Scope for a `_t` call site.
 * Priority: explicit `{ scope }` → withI18nScope stack → path[@location].
 */
export function resolveI18nScope(opts?: ResolveI18nScopeOptions): string {
  const manual = opts?.scope?.trim();
  if (manual) {
    return manual;
  }
  if (scopeStack.length > 0) {
    return scopeStack[scopeStack.length - 1];
  }
  return formatScope(opts?.path || '', opts?.location);
}

/** Test-only: reset scope stack between cases. */
export function __resetI18nScopeStackForTests(): void {
  scopeStack.length = 0;
}
